feature/carousel-example

DaveBitter opened the pull request:
**Currently, the boilerplate does not contain an extensive example component. This PR adds a carousel with some JavaScript logic and SCSS styles.** The module is written in a progressively enhanced manner.

Obviously, the code is opinionated. This will at least open the discussion on how modules 'need' to be written. I don't expect everybody to agree to a single way of building components, but I would like to agree on a set of standard rules.

Questions to the contributors:

- Do you have any remarks or improvements to the written JavaScript code? (styles, techniques...)
- Do you have any remarks or improvements to the written HTML code? (accessibility, structure...)
- Do you have any remarks or improvements to the written SCSS code? (fallbacks, structure...)
- Should I turn the hardcoded sizing values into variables? For now, I didn't for the sake of simplicity and overhead.

Let the discussions begin 🏁

peeke commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,143 @@
+  _hidePage(index) {
+    this._pages[index].setAttribute(constants.attributes.DATA_HIDDEN, 'true')
+  }
+
```
I think the navigation logic is a bit unclear at first glance.  Also, what if you add bullets and need to go from page 1 to page 3 for instance? 

I think you could split this in 3 methods: gotoPage, nextPage and previousPage 
```
gotoPage(index) {
    // Goto page logic
}
```
Passing the index to go to explicitly, you can completely remove the `getNextPage()` method, since the page to navigate to is already given. Keeping the index within the bounds allowed can be done here instead of in getNextPage. This would also play well with possible future bullet navigation.

```
nextPage() {
    gotoPage(this.state.indexCurrent + 1)
}

previousPage() {
    gotoPage(this.state.indexCurrent - 1)
}

```

peeke commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,143 @@
+  /**
+   * @param {Event} e - Click event emitted by the button
+   */
+  _handleNavigation(e) {
```
Since you already bind this to a buttonNext and buttonPrevious, I'd split this up in a `gotoNext` and `gotoPrevious`, which would allow you to scrap the logic to determine which direction you are navigation in.

peeke commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,143 @@
+  /**
+   * @param {Number} action - Amount to navigate
+   */
+  _navigate(action) {
```
If function took the index to goto, instead of an action, you could remove the need to store the previous index entirely:

```
gotoPage(index) {
   if (this.state.currentIndex === index) return
   this.hidePage(this.state.currentIndex)
   this.state.currentIndex = index
   this.showPage(this.state.currentIndex)
}
```

peeke commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,143 @@
+  /**
+   * @param {Number} action - Amount to navigate
+   * @returns {Number} - Index to navigate to
+   */
```
You can actually use the modulo operator for this: 
```
const loop = (value, length) => (value + length) % length

gotoPage(index) {
   index = loop(index, this.pages.length)
}
```
The reason you have to add the

DaveBitter replied:
Could you elaborate on this? I'm not quite sure what you mean.

peeke replied:
Of course, although maybe it's better not to use it in the example component if it's unclear.

The modulo operator is the % sign. It returns the remainder of an integer division. `a % b` means divide a by b to find a whole number, then return the remainder. For 5 % 4 that would be 1 for instance.

This effectively let's you loop your number around a fixed range. If your range is 4 and you try to go to slide 5 (index 4), `4 % 4` evaluates to 0, resetting the index to the beginning. In JavaScript a modulo on negative numbers works a bit different from other languages: `-1 % 4` evaluates to -1 instead of 3. That's why I added the range value first, to make the number positive. In the example above that would evaluate to: `(-1 + 4) % 4`, equaling 3. This then fixes looping the other way around as well.

Whenever I use this I usually tuck the logic away in a utility function, so the functionality is clear.

peeke replied:
This might only confuse, but for completeness sake: to be really safe, the function above should actually be 
```
const loop = (value, length) => (value % length + length) % length
```
Otherwise `loop(-5, 4)` would break (-5 + 4 equals -1, which is still negative). That case would never occur in the carousel example though.

YejunWu168 commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,143 @@
+  _addEventListeners() {
+    this._buttonPrev.addEventListener(
+      'click',
+      this._handleNavigation.bind(this)
```
You can use arrow functions on class properties to auto bind handleNavigation.

DaveBitter replied:
We got four ways of doing this. Calling bind, anonymous function to bind, bind in constructor and function call at the method. We have to agree to one, which is going to be a fun one.

colin-aarts replied:
5 ways, you forgot my weapon of choice: decorators (e.g. https://github.com/simonhoss/autobind-decorator).

DaveBitter replied:
Hahaha, well then we really need to come up with a single one. Not sure about installing a package though :).

peeke replied:
I don't have a preference, only thing to add:
If you bind your functions, might as well bind them in the constructor, so you have a reference to the function if you would like to remove the event listener later. Which reminds me: we should probably add an unload function as well, right?

waarissyb commented on src/components/carousel/_carousel.scss:
```diff
@@ -0,0 +1,42 @@
+    list-style-type: none;
+
+    .has-js & {
+      overflow-x: visible;
```
Having `overflow-x: visible;` seems counter-intuitive when you’re hiding the items with JavaScript. The value `initial` would be better in this case, but it seems good to have IE support.

DaveBitter commented:
I've squashed my commits into one, but wasn't able to figure out how to move the PR branch to a different repo. I'm not sure this is possible.

As per @waarissyb his feedback, I updated the PR to load images. I wasn't able to find the PR you mentioned from @awolkers. I've added placeholder images myself and used the `picture` element.

I've also updated the CSS to match @waarissyb his suggestions and I made us of the `aria-hidden` attribute for the sake of accessibility. @Thomas-Machielsen Could you do a review/addition for the accessibility part?

Finally, I've added an unload method, as per @peeke his suggestion.

Timvdv commented:
Test comment to check if the Slack Github integration is running!

DaveBitter commented:
I've resolved the final piece of feedback by @waarissyb . If there are no further changes required, I propose we merge this pull request so we have a solid example for people to view for our best practices.

DaveBitter commented:
@waarissyb can you review the changes, approve and merge if satisfied?

DaveBitter commented:
@waarissyb I added the Mocha test, as suggested. Can you update your 'changes requested' if satisfied?

DaveBitter commented:
Great! I don't have the almighty merge powers, so can somebody merge it who has?

colin-aarts commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,142 @@
+    this._init()
+  }
+
+  _defaultOptions = {}
```
This should be a `static` property, otherwise `Carousel.__defaultOptions` won't work.

colin-aarts commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,142 @@
+class Carousel {
+  /**
+   * @param {HTMLElement} element - The HTMLElement this module is constructed upon
+   * @param {Object} options - ConditionerJS's merged options
```
`@classdesc` could be used to summarise the class's purpose.

colin-aarts replied:
Also, conditioner? 😂

colin-aarts commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,142 @@
+   * Cache all selectors that are needed
+   */
+  _cacheSelectors() {
+    this._pages = Array.from(
```
Maybe a bit nitpicky, but if using spreading for the options above (`{ ...Carousel._defaultOptions, ...options }`) I think it'd be more consistent to use a spread here as well instead of `Array.from`. Or change the object spread to `Object.assign`.

colin-aarts commented on src/components/carousel/Carousel.js:
```diff
@@ -0,0 +1,142 @@
+// Constants
+const constants = {
```
These should be (default) options so that they can be overridden through the passed-in options if needed. It's free flexibility without a disadvantage. As a bonus, this also allows you to easily document them through `@prop`.

DaveBitter commented:
Thanks @colin-aarts ! Solid points. I've resolved the issues/remarks you had.
